rest-client/src/main/java/io/apicurio/registry/client/RegistryRestClientImpl.java

```diff
@@ -79,31 +112,84 @@
         initServices(retrofit);
     }
 
-    RegistryRestClientImpl(String baseUrl, Map<String, String> headers) {
-        if (!baseUrl.endsWith("/")) {
-            baseUrl += "/";
-        }
+    private static OkHttpClient createHttpClientWithConfig(String baseUrl, Map<String, Object> configs) {
+        OkHttpClient.Builder okHttpClientBuilder = new OkHttpClient.Builder();
+        okHttpClientBuilder = addHeaders(okHttpClientBuilder, baseUrl, configs);
+        okHttpClientBuilder = addSSL(okHttpClientBuilder, configs);
+        return okHttpClientBuilder.build();
+    }
 
-        final OkHttpClient okHttpClient = createWithHeaders(headers);
+    private static OkHttpClient.Builder addHeaders(OkHttpClient.Builder okHttpClientBuilder, String baseUrl, Map<String, Object> configs) {
 
-        this.retrofit = new Retrofit.Builder()
-                .client(okHttpClient)
-                .addConverterFactory(JacksonConverterFactory.create())
-                .baseUrl(baseUrl)
-                .build();
+        Map<String, String> requestHeaders = configs.entrySet().stream()
+            .filter(map -> map.getKey().startsWith(REGISTRY_REQUEST_HEADERS_PREFIX))
+            .collect(Collectors.toMap(map -> map.getKey()
+                .replace(REGISTRY_REQUEST_HEADERS_PREFIX, ""), map -> map.getValue().toString()));
 
-        this.requestHandler = new RequestHandler();
+        if(!requestHeaders.containsKey("Authorization")) {
+            // Check if url includes user/password
+            // and add auth header if it does
+            HttpUrl url = HttpUrl.parse(baseUrl);
+            String user = url.encodedUsername();
+            String pwd = url.encodedPassword();
+            if (user != null && !user.isEmpty()) {
+                String credentials = Credentials.basic(user, pwd);
+                requestHeaders.put("Authorization", credentials);
+            }
+        }
 
-        initServices(retrofit);
+        if(!requestHeaders.isEmpty()) {
+            final Interceptor headersInterceptor = new HeadersInterceptor(requestHeaders);
+            return okHttpClientBuilder.addInterceptor(headersInterceptor);
+        } else {
+            return okHttpClientBuilder;
+        }
     }
 
-    private static OkHttpClient createWithHeaders(Map<String, String> headers) {
-
-        final Interceptor headersInterceptor = new HeadersInterceptor(headers);
-
-        return new OkHttpClient.Builder()
-                .addInterceptor(headersInterceptor)
-                .build();
+    private static OkHttpClient.Builder addSSL(OkHttpClient.Builder okHttpClientBuilder, Map<String, Object> configs) {
+
+        try {
+            KeyManager[] keyManagers = null;
+            TrustManager[] trustManagers = null;
+
+            if (configs.containsKey(REGISTRY_REQUEST_KEYSTORE_LOCATION)) {
```

Comment:
Maybe these checks could be extracted to separated methods?

Reply:
Good idea - I have refactored in my latest commit 👍